### PR TITLE
Allow new users to be registered via the admin API even if the monthly active user limit has been reached

### PR DESCRIPTION
**Pull request by dklimpel**
Fixes: #7175
If MAU limit is reached, you can add user with `GET /_synapse/admin/v1/register`. With `PUT /_synapse/admin/v2/users/<user_id>` you can only add users if admin user was already active before limit was reached. Admin users seem to be MAU limited.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

File: changelog.d/7263.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Create user with admin API if MAU limit is reached.
```
Comment (richvdh) [inline changelog.d/7263.bugfix:1]: this sounds like every time the MAU limit is reached, we will create a new user!

```suggestion
Allow new users to be registered via the admin API even if the monthly active user limit has been reached.
```

File: synapse/handlers/register.py [hunk 2]
```diff
@@ -163,7 +163,9 @@ def register_user(
         """
         yield self.check_registration_ratelimit(address)
 
-        yield self.auth.check_auth_blocking(threepid=threepid)
+        # do not check_auth_blocking if the call is coming through the Admin API
+        if address:
```
Comment (richvdh) [inline synapse/handlers/register.py:167]: this is not reliable. There are other code paths which will set `address=None`.

I think the easiest solution is to add yet another parameter to `register_user` indicating whether the request came from the admin API, though I have to say that refactoring it into smaller functions is also compelling...

File: synapse/handlers/register.py [hunk 3]
```diff
@@ -156,14 +157,17 @@ def register_user(
               will be set to this. Defaults to 'localpart'.
             address (str|None): the IP address used to perform the registration.
             bind_emails (List[str]): list of emails to bind to this account.
+            by_admin (bool): Whether this registration was made by an administrator.
```
Comment (richvdh) [inline synapse/handlers/register.py:160]:
```suggestion
            by_admin (bool): True if this registration is being made via the admin api, otherwise False.
```

File: tests/rest/admin/test_user.py [hunk 4]
```diff
@@ -320,6 +324,59 @@ def nonce():
...
+            handler.register_user(localpart="local_part"), ResourceLimitError
+        )
+
+        store.get_monthly_active_count = Mock(
+            return_value=defer.succeed(self.hs.config.max_mau_value + 1)
+        )
+        self.get_failure(
+            handler.register_user(localpart="local_part"), ResourceLimitError
+        )
+
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:352]: why do we do this twice?

File: tests/rest/admin/test_user.py [hunk 5]
```diff
@@ -320,6 +324,59 @@ def nonce():
...
+        self.hs.config.max_mau_value = 2
+        self.hs.config.mau_trial_days = 0
+
+        # Set MAU limit
+        store.get_monthly_active_count = Mock(
+            return_value=defer.succeed(self.hs.config.max_mau_value + 1)
+        )
+        self.get_failure(
+            handler.register_user(localpart="local_part"), ResourceLimitError
+        )
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:344]: can you please add a comment to say what this is meant to achieve?

File: tests/rest/admin/test_user.py [hunk 6]
```diff
@@ -344,7 +401,7 @@ def test_no_auth(self):
         self.render(request)
 
         self.assertEqual(401, int(channel.result["code"]), msg=channel.result["body"])
-        self.assertEqual("M_MISSING_TOKEN", channel.json_body["errcode"])
+        self.assertEqual(Codes.MISSING_TOKEN, channel.json_body["errcode"])
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:404]: I'm not entirely convinced that this is an improvement; in any case it doesn't seem relevant to the change at hand?

File: tests/rest/admin/test_user.py [hunk 7]
```diff
@@ -514,6 +572,109 @@ def test_create_user(self):
...
+
+        request, channel = self.make_request(
+            "PUT",
+            url,
+            access_token=self.admin_user_tok,
+            content=body.encode(encoding="utf_8"),
+        )
+        self.render(request)
+
+        self.assertEqual(500, int(channel.result["code"]), msg=channel.result["body"])
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:676]: help me out here: what is the difference between these tests and the ones further up at line 327 etc? what is the purpose of calling sync etc?

Perhaps some more comments to explain the test strategy and whwhat we're trying to do would be helpful.

File: tests/rest/admin/test_user.py [hunk 8]
```diff
@@ -514,6 +572,109 @@ def test_create_user(self):
...
+        _do_sync_for_user(self.admin_user_tok)
+
+        self.store.get_monthly_active_count = Mock(
+            return_value=defer.succeed(self.hs.config.max_mau_value + 1)
+        )
+
+        # Set MAU limit
+        self.get_failure(
+            handler.register_user(localpart="local_part"), ResourceLimitError
+        )
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:607]: I don't think this is setting a MAU limit.

File: tests/rest/admin/test_user.py [hunk 9]
```diff
@@ -514,6 +572,109 @@ def test_create_user(self):
...
+
+        self.store.get_monthly_active_count = Mock(
+            return_value=defer.succeed(self.hs.config.max_mau_value + 1)
+        )
+        self.get_failure(
+            handler.register_user(localpart="local_part"), ResourceLimitError
+        )
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:614]: again: why do we do exctly the same thing twice?

File: tests/rest/admin/test_user.py [hunk 10]
```diff
@@ -514,6 +572,109 @@ def test_create_user(self):
...
+            request, channel = self.make_request("GET", "/sync", access_token=token)
+            self.render(request)
+
+            if channel.code != 200:
+                raise HttpResponseException(
+                    channel.code, channel.result["reason"], channel.result["body"]
+                ).to_synapse_error()
+
+        # Sync to set admin user to active
+        _do_sync_for_user(self.admin_user_tok)
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:598]: why do we do this as well as the mock of `get_monthly_active_count`?

File: tests/rest/admin/test_user.py (hunk 6 above)
Comment (dklimpel) [reply tests/rest/admin/test_user.py:404]: Yes, you're right. It is not an improvement. It is the try to use consistently error codes as constant in all tests.

File: tests/rest/admin/test_user.py (hunk 4 above)
Comment (dklimpel) [reply tests/rest/admin/test_user.py:352]: It is my mistake.

File: tests/rest/admin/test_user.py (hunk 7 above)
Comment (dklimpel) [reply tests/rest/admin/test_user.py]: I have added more comments.
The difference is the `sync`.
If admin did a sync before mau limit is reached, the admin can send `POST` requests and create new users.
If admin did not a `sync` before limit is reached, the admin can not create new users, because the admin can not send `POST` requests. The admin is limited like a normal user.

**Comment by clokep** [comment]
@dklimpel Is this ready for review again or are you still handling the review comments?

**Comment by dklimpel** [comment]
@clokep It is ready for review.

File: tests/rest/admin/test_user.py [hunk 11]
```diff
@@ -320,6 +324,55 @@ def nonce():
...
+        Register user with Admin API if MAU limit is reached.
+        """
+        handler = self.hs.get_registration_handler()
+        store = self.hs.get_datastore()
+        auth_blocking = self.hs.get_auth()._auth_blocking
+
+        # Configure MAU limit
+        auth_blocking._limit_usage_by_mau = True
+        auth_blocking._max_mau_value = 2
+        auth_blocking._mau_trial_days = 0
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:338]: better to do this with `@override_config`, rather than poking into the guts of the auth_blocking object

or better yet: override `default_config`.

See also the docstring on `HomeserverTestCase`.

File: tests/rest/admin/test_user.py [hunk 12]
```diff
@@ -320,6 +324,55 @@ def nonce():
         self.assertEqual(400, int(channel.result["code"]), msg=channel.result["body"])
         self.assertEqual("Invalid user type", channel.json_body["error"])
 
+    def test_register_mau_limit_reached(self):
+        """
+        Register user with Admin API if MAU limit is reached.
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:329]:
```suggestion
        Check we can register a user via the shared secret registration API even if the MAU limit is reached.
```

File: tests/rest/admin/test_user.py [hunk 13]
```diff
@@ -320,6 +324,55 @@ def nonce():
...
+        auth_blocking._mau_trial_days = 0
+
+        # Set monthly active users to the limit
+        store.get_monthly_active_count = Mock(
+            return_value=defer.succeed(auth_blocking._max_mau_value)
+        )
+        # The registration of a new user fails due to the limit
+        self.get_failure(
+            handler.register_user(localpart="local_part"), ResourceLimitError
+        )
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:347]: why do we do this? to check that the auth blocking is working as expected? please say so in a comment if so

File: tests/rest/admin/test_user.py (hunk 6 above)
Comment (richvdh) [reply tests/rest/admin/test_user.py:400]: if it is not an improvement, we should not do it.

it seems to me that the test is more thorough if we check the actual result. If we use the constant, we wouldn't detect a typo in the definition of the constant.

File: tests/rest/admin/test_user.py [hunk 14]
```diff
@@ -419,7 +473,7 @@ def test_user_does_not_exist(self):
         self.render(request)
 
         self.assertEqual(404, channel.code, msg=channel.json_body)
-        self.assertEqual("M_NOT_FOUND", channel.json_body["errcode"])
+        self.assertEqual(Codes.NOT_FOUND, channel.json_body["errcode"])
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:476]: revert this please

File: tests/rest/admin/test_user.py [hunk 15]
```diff
@@ -516,6 +570,103 @@ def test_create_user(self):
...
+        """
+        self.hs.config.registration_shared_secret = None
+
+        handler = self.hs.get_registration_handler()
+        auth_blocking = self.hs.get_auth()._auth_blocking
+
+        # Configure MAU limit
+        auth_blocking._limit_usage_by_mau = True
+        auth_blocking._max_mau_value = 2
+        auth_blocking._mau_trial_days = 0
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:586]: again, do this elsewhere please.

File: tests/rest/admin/test_user.py [hunk 16]
```diff
@@ -516,6 +570,103 @@ def test_create_user(self):
...
+            access_token=self.admin_user_tok,
+            content=body.encode(encoding="utf_8"),
+        )
+        self.render(request)
+
+        self.assertEqual(201, int(channel.result["code"]), msg=channel.result["body"])
+        self.assertEqual("@bob:test", channel.json_body["name"])
+        self.assertEqual(False, channel.json_body["admin"])
+
+    def test_create_user_mau_limit_reached_passiv_admin(self):
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:627]:
```suggestion
    def test_create_user_mau_limit_reached_passive_admin(self):
```

File: tests/rest/admin/test_user.py [hunk 17]
```diff
@@ -516,6 +570,103 @@ def test_create_user(self):
         self.assertEqual(False, channel.json_body["is_guest"])
         self.assertEqual(False, channel.json_body["deactivated"])
 
+    def test_create_user_mau_limit_reached_active_admin(self):
+        """
+        Check that a new regular user is created successfully if MAU limit is reached.
+        Admin user was active before creating user.
+        """
+        self.hs.config.registration_shared_secret = None
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:578]: and this

File: tests/rest/admin/test_user.py [hunk 18]
```diff
@@ -516,6 +570,103 @@ def test_create_user(self):
         self.assertEqual(False, channel.json_body["is_guest"])
         self.assertEqual(False, channel.json_body["deactivated"])
 
+    def test_create_user_mau_limit_reached_active_admin(self):
+        """
+        Check that a new regular user is created successfully if MAU limit is reached.
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:575]:
```suggestion
        Check that an admin can register a new user via the admin API even if the MAU limit is reached.
```

File: tests/rest/admin/test_user.py [hunk 19]
```diff
@@ -516,6 +570,103 @@ def test_create_user(self):
...
+        # before limit of monthly active users is reached
+        request, channel = self.make_request(
+            "GET", "/sync", access_token=self.admin_user_tok
+        )
+        self.render(request)
+
+        if channel.code != 200:
+            raise HttpResponseException(
+                channel.code, channel.result["reason"], channel.result["body"]
+            ).to_synapse_error()
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:598]: what is the point of calling `to_synapse_error` here?

File: tests/rest/admin/test_user.py [hunk 20]
```diff
@@ -516,6 +570,103 @@ def test_create_user(self):
...
+        )
+        self.render(request)
+
+        self.assertEqual(201, int(channel.result["code"]), msg=channel.result["body"])
+        self.assertEqual("@bob:test", channel.json_body["name"])
+        self.assertEqual(False, channel.json_body["admin"])
+
+    def test_create_user_mau_limit_reached_passiv_admin(self):
+        """
+        Try to create a new regular user if MAU limit is reached.
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:629]:
```suggestion
         Check that an admin can register a new user via the admin API even if the MAU limit is reached.
```

File: tests/rest/admin/test_user.py [hunk 21]
```diff
@@ -516,6 +570,103 @@ def test_create_user(self):
...
+        Admin user was not active before creating user.
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:630]: I'm still a bit baffled by why we need this as well as the previous test. Is it really likely that an inactive admin could be blocked from creating a new user?

Comment (dklimpel) [reply tests/rest/admin/test_user.py:631]: The test is presented for completeness. It can be left out. We deal with mau. And it would be possible that an active and passive admin/user get different results.

File: tests/rest/admin/test_user.py [hunk 22]
```diff
@@ -368,8 +417,14 @@ class UserRestTestCase(unittest.HomeserverTestCase):
...
         synapse.rest.admin.register_servlets,
         login.register_servlets,
+        sync.register_servlets,
     ]
 
+    def default_config(self):
+        config = super().default_config()
+        config.update({"registration_shared_secret": None})
+        return config
+
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:427]: this is redundant afaict
```suggestion
```

**Comment by richvdh** [comment]
thanks again for your work on this.
